lambdas/src/apis/collections/controllers/collections.ts

```diff
@@ -150,46 +200,96 @@ function buildUrn(protocol: string, contract: string, option: string): string {
 }
 
-async function internalContents(
-  client: SmartContentClient,
-  res: Response,
-  urn: string,
-  selector: (metadata: WearableMetadata) => string | undefined
-): Promise<void> {
-  try {
-    const entity = await fetchEntity(client, urn)
-    if (entity) {
-      const wearableMetadata: WearableMetadata = entity.metadata
-      const hash = findHashForFile(entity, selector(wearableMetadata))
-      if (hash) {
-        const headers: Map<string, string> = await client.pipeContent(hash, res as any as ReadableStream<Uint8Array>)
-        headers.forEach((value: string, key: string) => {
-          res.setHeader(key, value)
-        })
-      } else {
-        res.status(404).send()
-      }
-    }
-  } catch (e) {
-    res.status(500).send(e.message)
-  }
+function validateSize(size: string = DEFAULT_IMAGE_SIZE): size is ValidSize {
+  if (!isValidSize(size)) throw new ServiceError('Invalid size')
+  return true
+}
+
+function getFileHash(entity: Entity, fileName?: string): string {
+  const hash = findHashForFile(entity, fileName)
+  if (!hash) throw new ServiceError(`Hash not found for file ${fileName}`, 404)
+
+  return hash
 }
 
-async function fetchEntity(client: SmartContentClient, urn: string): Promise<Entity | undefined> {
+async function fetchEntity(client: SmartContentClient, urn: string): Promise<Entity> {
   const entities: Entity[] = await client.fetchEntitiesByPointers(EntityType.WEARABLE, [urn])
-  return entities && entities.length > 0 && entities[0].metadata ? entities[0] : undefined
+  if (!(entities && entities.length > 0 && entities[0].metadata)) throw new ServiceError('Entity not found', 404)
+
+  return entities[0]
 }
 
-export function getBodyShapes(representations: WearableMetadataRepresentation[]) {
+function getBodyShapes(representations: WearableMetadataRepresentation[]): string[] {
   const bodyShapes = new Set<WearableBodyShape>()
   for (const representation of representations) {
     for (const bodyShape of representation.bodyShapes) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
       bodyShapes.add(bodyShape.split(':').pop()!)
     }
   }
   return Array.from(bodyShapes)
 }
 
+function isValidSize(size: string): size is ValidSize {
+  return sizes[size] !== undefined
+}
+
+function getImagePath(root: string, imageRequest: ImageRequest): string {
+  return root + `/` + getImageRequestId(imageRequest) + '.png'
+}
+
+function getImageRequestId({ urn, hash, size, rarity }: ImageRequest): string {
+  return `${urn}-${hash}-${size}` + (rarity ? `-${rarity}` : '')
+}
+
+async function getImage(
+  client: SmartContentClient,
+  rootStorageLocation: string,
+  imageRequest: ImageRequest
+): Promise<Buffer> {
+  const path = getImagePath(rootStorageLocation, imageRequest)
+
+  // Check if the image is already in the cache, otherwise build and store it
+  fs.existsSync(path) || (await buildImage(client, rootStorageLocation, imageRequest))
```

Comment:
Would it make sense to have a metric that checks the number of images built? (+ The space on disk that they are using)